### DISCUSSION TITLE
OpenWRT 24.10

### DISCUSSION THREAD

**Discussion by ZerBea** [ZerBea/hcxdumptool#478] [General]
OpenWRT 24.10 is available now:
```
$ git branch -a
  master
  openwrt-23.05
* openwrt-24.10
  remotes/origin/HEAD -> origin/master
  remotes/origin/dgs-1210-28p-luizluca
  remotes/origin/lede-17.01
  remotes/origin/main
  remotes/origin/master
  remotes/origin/openwrt-18.06
  remotes/origin/openwrt-19.07
  remotes/origin/openwrt-21.02
  remotes/origin/openwrt-22.03
  remotes/origin/openwrt-23.05
  remotes/origin/openwrt-24.10
```

Initial tests have shown that 24.10 is working as fine.

Some notes:
The init process of some devices is slow.
Add "sleep 10" to an init.d script to make sure the device has been initialized before hcxdumptool/hcxlabtool starts.

Running a minimalist OpenWRT, it could be possible that a WiFi USB adapter has not been assigned during boot.
Add "sudo iw phy phyx interface add wlan0 type managed" to an init.d script to make sure the interface has been assigned (hcxdumptool/hcxlabtool does not hijack unassigned interfaces).

**Comment by ZerBea** [comment]
Update:
Build failed on latest git head (as of today):
https://github.com/openwrt/openwrt/issues/16905

**Reply by strasharo** [reply]
Damn, I was literally just about to start building it. :laughing: What are your general impressions, is it stable enough or it's better to stick with openwrt-23.05 for now?

**Comment by ZerBea** [comment]
I'm still testing 24.10 (compile last week). It is working fine and stable.

But there are some negative aspects:
OpenWRT Linux kernel 6.6.58 is running fine, but it looks like the rtl8xxxu (kmod-rtl8xxxu) is not the same as the Linux stock kernel driver.
While monitor mode on the Linux stock kernel driver is working fine, it doesn't work on the OpenWRT driver.

**Comment by ZerBea** [comment]
I switched to the master branch and it (mac80211) compiles fine:
```
 make[3] -C package/network/config/wifi-scripts compile
 make[3] -C package/kernel/mac80211 compile
 make[3] -C package/libs/ncurses compile
```

**Comment by ZerBea** [comment]
Looks like the newly compiled image (from master branch) is working as expected. Now show stopper yet.

**Comment by ZerBea** [comment]
After some initial hinds, OpenWRT 24.10 performs well (as of today).
Initial tests, running a cheap Realtek device [TL-WN722N v2/3](https://www.reichelt.de/de/en/wireless-lan-usb-adapter-150-mbit-s-detachable-aerial-tplink-tl-wn722n-p125227.html?GROUPID=5839&START=0&OFFSET=16&SID=956ce594175db3f54c0d5eaf789be182ca2835f6e8aa003d02818&LANGUAGE=EN&&r=1)
are impressive.

**Comment by ZerBea** [comment]
To disable the red PWR LED and/or green ACT LED of a Raspberry Pi add this 2 lines to an init.d script:
```
if [ -f "/sys/class/leds/PWR/brightness" ]; then echo 0 > /sys/class/leds/PWR/brightness; fi
if [ -f "/sys/class/leds/ACT/brightness" ]; then echo 0 > /sys/class/leds/ACT/brightness; fi
```
Estimated power savings about 5 mA.

**Comment by strasharo** [comment]
Hey @ZerBea, just noticed the following line in lessons_learned.txt:
"sharing the interface (e.g. iw phy phy1 interface add mon0 type monitor) is counterproductive"

By default in Openwrt I just get phy0 without an interface name assigned to it. What would be the best way to use it in this case?

**Reply by ZerBea** [reply]
OpenWRT does not assign interfaces and hcxdumptool/hcxlabtool does not hijack unassigned interfaces. A solution could be to assign the interface during boot. 
A simple script, placed in /etc/init.d should do the trick:

```
#!/bin/sh /etc/rc.common
START=99

boot()
	{
	/bin/sleep 10
	/usr/sbin/iw phy phy0 interface add wlan0 type managed
	}
```

Now hcxdumptool/hcxlabtool sees an assigned managed interface and can set monitor mode.
sleep 10 is useful to make sure OpenWRT got a list of all phy interfaces before we assign it.
https://github.com/ZerBea/hcxdumptool/discussions/478#discussion-7412789

**Reply by strasharo** [reply]
Thank you.

**Comment by ZerBea** [comment]
Full rtw88 USB support has been added to OpenWRT 24.10 (latest git head);
```
package/firmware/linux-firmware/realtek.mk                   |   14 +
 ...001-v6.12-wifi-rtw88-Set-efuse-ext_lna_5g-fix-typo.patch |   30 +
 ...fi-rtw88-usb-Support-USB-3-with-RTL8822CU-RTL8822B.patch |  284 +++
 ...fi-rtw88-debugfs-support-multiple-adapters-debuggi.patch |  454 ++++
 .../rtl/005-v6.12-wifi-rtw88-select-WANT_DEV_COREDUMP.patch |   25 +
 ...fi-rtw88-8822c-Parse-channel-from-IE-to-correct-in.patch |  151 ++
 ...fi-rtw88-usb-Init-RX-burst-length-according-to-USB.patch |   73 +
 ...fi-rtw88-usb-Update-the-RX-stats-after-every-frame.patch |   31 +
 .../012-v6.12-wifi-rtw88-usb-Support-RX-aggregation.patch   |  118 +
 ...fi-rtw88-Enable-USB-RX-aggregation-for-8822c-8822b.patch |  168 ++
 ...fi-rtw88-assign-mac_id-for-vif-sta-and-update-to-T.patch |  231 ++
 ...-v6.13-wifi-rtw88-Constify-some-arrays-and-structs.patch |  374 ++++
 ...fi-rtw88-Parse-the-RX-descriptor-with-a-single-fun.patch |  560 +++++
 ...12-wifi-rtw88-Fix-the-RX-aggregation-in-USB-3-mode.patch |   70 +
 ...fi-rtw88-Refactor-looping-in-rtw_phy_store_tx_powe.patch |   34 +
 ...fi-rtw88-Report-the-signal-strength-only-if-it-s-k.patch |   39 +
 ...fi-rtw88-Add-some-definitions-for-RTL8821AU-RTL881.patch |  490 +++++
 ...ifi-rtw88-Dump-the-HW-features-only-for-some-chips.patch |   93 +
 ....13-wifi-rtw88-Allow-different-C2H-RA-report-sizes.patch |  175 ++
 ...fi-rtw88-Extend-the-init-table-parsing-for-RTL8812.patch |  165 ++
 ...fi-rtw88-Allow-rtw_chip_info.ltecoex_addr-to-be-NU.patch |   58 +
 ...fi-rtw88-Let-each-driver-control-the-power-on-off-.patch |  272 +++
 ...fi-rtw88-Enable-data-rate-fallback-for-older-chips.patch |  194 ++
 ...-v6.13-wifi-rtw88-Make-txagc_remnant_ofdm-an-array.patch |   85 +
 ...3-wifi-rtw88-Support-TX-page-sizes-bigger-than-128.patch |  115 +
 ...ifi-rtw88-Move-pwr_track_tbl-to-struct-rtw_rfe_def.patch |  297 +++
 ...fi-rtw88-usb-Set-pkt_info.ls-for-the-reserved-page.patch |   27 +
 ...fi-rtw88-Detect-beacon-loss-with-chips-other-than-.patch |   63 +
 ...wifi-rtw88-coex-Support-chips-without-a-scoreboard.patch |   66 +
 ...wifi-rtw88-8821a-Regularly-ask-for-BT-info-updates.patch |   67 +
 .../038-v6.13-wifi-rtw88-8812a-Mitigate-beacon-loss.patch   |   37 +
 .../rtl/039-v6.13-wifi-rtw88-Add-rtw8812a_table.-c-h.patch  | 2862 +++++++++++++++++++++++++
 .../rtl/040-v6.13-wifi-rtw88-Add-rtw8821a_table.-c-h.patch  | 2395 +++++++++++++++++++++
 .../rtl/041-v6.13-wifi-rtw88-Add-rtw88xxa.-c-h.patch        | 2187 +++++++++++++++++++
 .../rtl/042-v6.13-wifi-rtw88-Add-rtw8821a.-c-h.patch        | 1230 +++++++++++
 .../rtl/043-v6.13-wifi-rtw88-Add-rtw8812a.-c-h.patch        | 1135 ++++++++++
 ...4-v6.13-wifi-rtw88-Add-rtw8821au.c-and-rtw8812au.c.patch |   80 +
 ...fi-rtw88-Enable-the-new-RTL8821AU-RTL8812AU-driver.patch |  153 ++
 ...fi-rtw88-8821au-Add-additional-devices-to-the-USB_.patch |   77 +
 .../rtl/047-wifi-rtw88-8812au-Add-more-device-IDs.patch     |   91 +
 .../048-wifi-rtw88-usb-Support-USB-3-with-RTL8812AU.patch   |   69 +
 ...fi-rtw88-usb-Enable-RX-aggregation-for-8821au-8812.patch |   65 +
 ...fi-rtlwifi-rtl8821ae-phy-restore-removed-code-to-f.patch |   32 +
 ...51-wifi-rtw88-Add-additional-USB-IDs-for-RTL8812BU.patch |   31 +
```

Initial test is here:
https://github.com/ZerBea/hcxdumptool/discussions/361#discussioncomment-11531030

**Reply by strasharo** [reply]
Any idea which would be the appropriate package to select for AWUS036AC (RTL8812AU)?
I'm running it on Arch using the rtw88 driver from git, but I'm not sure which one matches it from these in the OpenWRT menuconfig:

![image](https://github.com/user-attachments/assets/ffddb687-20e3-40b3-8238-6712ba99de9c)

**Comment by ZerBea** [comment]
You need to do a git pull to get the latest drivers (don't forget to update you .config):
![kmod](https://github.com/user-attachments/assets/94352c13-984e-404c-8640-1d993589401c)
Than select kmod-rtw88-8812au.

BTW:
According  @morrownr rtl8814au is in development but not ready at this time:
https://github.com/morrownr/USB-WiFi/issues/549

**Reply by strasharo** [reply]
Guess something was stale on my side, just switched branches to master and then back to openwrt-24.10 and it popped. Thanks a lot.

**Reply by ZerBea** [reply]
You're welcome.

BTW:
No need to switch to master and back.
set feeds to 24.10 (last one is my local hcxlabtool feed);
```
src-git packages https://git.openwrt.org/feed/packages.git;openwrt-24.10
src-git luci https://git.openwrt.org/project/luci.git;openwrt-24.10
src-git routing https://git.openwrt.org/feed/routing.git;openwrt-24.10
src-git telephony https://git.openwrt.org/feed/telephony.git;openwrt-24.10
src-link mypackages /home/zerobeat/Source/raspberry/openwrt_packages/
```
do make clean
update and install feeds
```
./scripts/feeds clean
./scripts/feeds update -a
./scripts/feeds install -a
```
do make menuconfig
do make -j $(($(nproc)+1))

**Comment by ZerBea** [comment]
The new (updated) drivers rtl8xxxu and rtw88 usb are real game changers since they joined the bandwagon.
Running hcxlabtool/hcxdumptool, almost every tested (cheap) rtl8xxxu or rtw88 usb device is performing better than (expensive) mt76 devices.

**Comment by morrownr** [comment]
@ZerBea @strasharo 

> Full rtw88 USB support has been added to OpenWRT 24.10 (latest git head)

I'm glad this is happening quickly. I did not expect it to happen this quickly. One thing to be aware of is that not all patches are in mainline yet so are likely not in OpenWRT. Examples: USB3 detect and autoswitch support for 8812bu just landed in kernel 6.13 rc2. USB3 detect and autoswitch for 8812au is scheduled to go into kernel 6.14. There are more. The amount of patches necessary to get USB support up to speed in rtw88 was large and took over a year.

It appears that I need to update some of my documentation at github.com/morrownr/USB-WiFi . It would help me if I see reports on the quality of the rtw88 support. Thanks.

**Reply by ZerBea** [reply]
USB3 is still a mess due to massive interferes. The same applies to WiFi/BT devices.
Running hcxlabtool/hcxdumptool I strongly recommend to use Realtek bgn devices (rtl8xxxu) first choice, directly followed by USB2 ac devices (rtw88 usb).
Please note that hcxlabtool/hcxdumptool is hunting for AUTHENTICATIONs. Almost they are done running slowest rate and smallest bandwidth.
5, 6, 7GHz is not mandatory, because CLIENTs probing the entire frequency range. Please take a look (Wireshark) at the radiotap header of a CLIENTs PROBEREQUEST that is looking for a 5GHz AP. It will do that on all bands - and hcxlabtool/hcxdumptool responds (on 2.4GHz)! It is merely a question of the IE-TAG of hcxdumptools/hcxlabtools PROBERESONSE.

**Reply by ZerBea** [reply]
This is all you need:
![pi](https://github.com/user-attachments/assets/d823de14-2c71-4575-95a4-7e8ca209242a)

Left hand side:
Raspberry Pi Zero (1) & TP-Link TL-WN823N (rtl8xxxu)
Right hand side:
Raspberry Pi Zero (1) & TP-Link Archer T2UV3 (rtw88 usb)
Bottom:
Powerbank

**Reply by ZerBea** [reply]
Or
![pi_b](https://github.com/user-attachments/assets/d0930873-9bdf-4a61-9161-dad645346a31)
Raspberry Pi Zero (1) & TP-Link TL-WN822N (rtl8xxxu)

**Reply by ZerBea** [reply]
Or, if you prefer an external antenna:
![pi_c](https://github.com/user-attachments/assets/986ee662-87c2-4680-8bf7-6ad319480f86)
Raspberry Pi Zero (1) & EW-7811USC  (rtw88 usb)

**Reply by ZerBea** [reply]
All penetration systems (RPI + WiFi adapter + USB cables + power bank) mentioned above are cheaper than a single ALFA AWUS036AXML - and in view of the current driver/firmware situation, they are performing much better!

**Reply by ZerBea** [reply]
@morrownr 
basic driver tests are here (as soon as I can put my hands on a device)
penetration test results only pgp encrypted

**Comment by morrownr** [comment]
@ZerBea 

> USB3 is still a mess due to massive interferes. The same applies to WiFi/BT devices.

I agree completely. Knowledge of the problems that USB3 creates is not widespread. Even some adapter makers are not aware of it. There are certain guidelines that I emphasize on my USB-WiFi site:

- Buy only single-function adapters (no onboard bluetooth support)
- Buy only single-state adapters (no onboard storage for a Windows driver)

My Plug and Play List only shows adapters that meet the above criteria. I probably need to go further and explain in detail the interference problems. Even using a modern dual or tri band adapter in band 1 (2.4 GHz) can be an issue and I recommend that users in this situation only use USB2.

**Reply by ZerBea** [reply]
At the moment I'm wrestling with the rtl8xxxu driver and a TP-LINK TL-WN8200ND (ID 2357:0126 TP-Link 802.11n NIC).
The device is EOL and they sell it out for a few EURO's.
```
[42361.010397] usb 1-9.2: New USB device found, idVendor=2357, idProduct=0126, bcdDevice= 2.00
[42361.010405] usb 1-9.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[42361.010408] usb 1-9.2: Product: 802.11n NIC
[42361.010410] usb 1-9.2: Manufacturer: Realtek
[42361.010412] usb 1-9.2: SerialNumber: 5091E3C3A656
[42361.047423] usb 1-9.2: This Realtek USB WiFi dongle (0x2357:0x0126) is untested!
[42361.047429] usb 1-9.2: Please report results to Jes.Sorensen@gmail.com
[42365.445651] usb 1-9.2: Dumping efuse for RTL8192EU (0x200 bytes):
[42365.445658] 00000000: 29 81 00 7c 01 40 03 00 70 34 04 50 14 00 00 00  )..|.@..p4.P....
[42365.445660] 00000010: 21 21 21 22 22 22 28 28 28 29 29 f0 00 ef ff ff  !!!"""((()).....
[42365.445661] 00000020: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445662] 00000030: ff ff ff ff ff ff ff ff ff ff 2d 2d 2d 2d 2d 2d  ..........------
[42365.445663] 00000040: 2d 2d 2d 2d 2d f0 ef ef ff ff ff ff ff ff ff ff  -----...........
[42365.445663] 00000050: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445664] 00000060: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445665] 00000070: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445666] 00000080: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445667] 00000090: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445668] 000000a0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445668] 000000b0: ff ff ff ff ff ff ff ff a1 2f 1b 00 00 00 ff ff  ........./......
[42365.445669] 000000c0: ff 01 00 10 00 00 00 ff 00 00 31 ff ff ff ff ff  ..........1.....
[42365.445670] 000000d0: 57 23 26 01 e6 47 02 50 91 e3 c3 a6 56 09 03 52  W#&..G.P....V..R
[42365.445671] 000000e0: 65 61 6c 74 65 6b 0d 03 38 30 32 2e 31 31 6e 20  ealtek..802.11n 
[42365.445672] 000000f0: 4e 49 43 00 ff ff ff ff ff ff ff ff ff ff ff ff  NIC.............
[42365.445673] 00000100: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445674] 00000110: ff ff ff ff ff ff ff 0d 03 00 05 00 30 00 00 00  ............0...
[42365.445674] 00000120: 00 93 ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445675] 00000130: f6 a8 98 2d 03 92 98 00 fc 8c 00 11 9b 44 02 0a  ...-.........D..
[42365.445676] 00000140: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445677] 00000150: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445678] 00000160: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445679] 00000170: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445679] 00000180: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445680] 00000190: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445681] 000001a0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445682] 000001b0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445683] 000001c0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445684] 000001d0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445684] 000001e0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445685] 000001f0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
[42365.445686] usb 1-9.2: RTL8192EU rev B (SMIC) romver 0, 2T2R, TX queues 3, WiFi=1, BT=0, GPS=0, HI PA=0
[42365.445689] usb 1-9.2: RTL8192EU MAC: 50:91:e3:c3:a6:56
[42365.445690] usb 1-9.2: rtl8xxxu: Loading firmware rtlwifi/rtl8192eu_nic.bin
[42365.445863] usb 1-9.2: Firmware revision 35.7 (signature 0x92e1)
[42374.525737] rtl8xxxu 1-9.2:1.0 wlp22s0f0u9u2: renamed from wlan0
```

The USB id has been added to the driver. While the old hal driver is working as expected, rtl8xxxu doesn't.
 The green LED flashes during scan, but the device doesn't transmit:
```
$ sudo ip link set wlp22s0f0u9u2 up
$ sudo iw dev wlp22s0f0u9u2 scan
$
```

**Reply by ZerBea** [reply]
BTW:
hcxdumptool and hcxtools are part of OpenWRT:
![hcxdumptool](https://github.com/user-attachments/assets/de9d7208-9b84-4334-a2f8-7107bb9fc293)

and of course they are part of Arch Linux, too:
https://archlinux.org/packages/?q=hcx

**Reply by strasharo** [reply]
Packages in OpenWRT are slightly behind currently. Opened PR's to bump them to 6.3.5, however they've asked to include my name in the commits, so I closed them. :/

**Reply by ZerBea** [reply]
Unfortunately, this is the case in many distributions.
hcxlabtool/hcxdumptool and hcxtools are kind of a rolling release. The release cycle depends on daily analysis of [wpa-sec uploads](https://wpa-sec.stanev.org/), the [Linux kernel cycle](https://www.kernel.org/) and of course the release cycle of [hashcat](https://github.com/hashcat/hashcat) and [JtR](https://github.com/openwall/john).
Unfortunately most of the package maintainers are very busy and they can't follow this cycle.
Luckily all hcx tools  compile well on all distributions if the dependencies are met (libopenssl, zlib, libcurl and pkgconfig for hcxtools).
And they should work as expected if the system is updated (WiFi network drivers for hcxlabtool/hcxdumptool).

**Reply by strasharo** [reply]
I would gladly volunteer to do the maintenance of the Openwrt packages, however that real name requirement they have is an overkill for me. :smile:  Guess I'll go with local feeds instead.

**Reply by ZerBea** [reply]
Well, let's see how long it takes until OpenWRT receives full rtl8192fu support.
This device is really worth it.
![RPI_EW](https://github.com/user-attachments/assets/012ac0a2-4174-4214-a9ee-43036671c9e6)

**Comment by ZerBea** [comment]
Missing firmware rtl8192fufw.bin (e.g. Edimax EW-7722UTn V3)
https://github.com/openwrt/openwrt/pull/17286

Driver/device test is here:
https://github.com/ZerBea/hcxdumptool/discussions/361#discussioncomment-10428718
